intermediate_source/nvfuser_intro_tutorial.py

```diff
@@ -0,0 +1,647 @@
+# For when we’ll use TorchScript on this function, we decorate the
+# function with type information of the function parameters. This isn’t
+# always required, but it can often help to provide this information to
+# TorchScript because it is a strictly typed system. Since we have
+# PyTorch’s autograd system, we don’t need to explicitly define the
+# backwards pass.
+
+
+def composite_definition(
+    input1: torch.Tensor,
+    input2: torch.Tensor,
+    weight: torch.Tensor,
+    bias1: torch.Tensor,
+    bias2: torch.Tensor,
+    normalization_axis: int,
+    dropout_prob: float,
+) -> torch.Tensor:
+    bias1_out = input1 + bias1
+    dropout_out = F.dropout(bias1_out, dropout_prob, training=True)
+    norm_input = dropout_out + input2
+    norm_output = F.layer_norm(
+        norm_input, (input1.size(normalization_axis),), weight, bias2)
+    return norm_output
+
+######################################################################
+# Setup and Performance Metrics
+# ---------------------
+# Next we initialize some inputs, parameters, and a simulated gradient
+# output tensor for the backwards pass since we aren’t including a
+# loss function.
+#
+
+
+# Setup initial tensors and parameters
+input_size = [64, 128, 1024]
+device = "cuda"
+dtype = torch.float32
+
+# Create sample inputs
+input1 = torch.randn(*input_size, device=device,
+                     dtype=dtype, requires_grad=True)
+input2 = torch.rand_like(input1).requires_grad_()
+
+# Precompute a grad output tensor, for this example it's the same size
+# as the inputs
+grad_output = torch.rand_like(input1)
+
+# Randomly initialize the model parameters
+weight = torch.nn.Parameter(torch.randn(
+    input_size[2], dtype=dtype, device=device))
+bias1 = torch.nn.Parameter(torch.randn(
+    input_size[2], dtype=dtype, device=device))
+bias2 = torch.nn.Parameter(torch.randn(
+    input_size[2], dtype=dtype, device=device))
+
+parameters = [input1, input2, weight, bias1, bias2]
+
+######################################################################
+# To produce a baseline performance we will measure the speed of our
+# forward and backward passes in PyTorch’s default eager mode. To get
+# accurate and comparable measurements, we perform a few warm up
+# iterations. Then we time many iterations of the forward and backward
+# pass using performance counters combined with proper GPU
+# synchronization, then compute the average iterations per second.
+# It’s important to be very careful when measuring performance on the
+# GPU, as we want to remove any initialization costs and need
+# synchronization since it’s an asynchronous device. Since we will
+# measure many variations of this problem with and without nvFuser we
+# define a helper method called `profile_workload` and will use
+# `functool.partial` to concisely profile the workload.
+#
+
+# Utility to profile the workload
+def profile_workload(forward_func, grad_output, iteration_count=100):
```

Comment:
Can we add a "label" parameter that gets printed with the output (like "eager" or "torchscript")? That will make it easier to compare the outputs of workloads with different settings.

Reply:
Done!